### PR TITLE
Cleanup `MegaBlocks` Installation

### PR DESCRIPTION
**Pull request by eitanturok**
**Update 2024-07-25:**

1. We will use `stk` for now and not my fork. We'll take care of this in a separate PR.
2. We will update the repo author and email in a separate PR.
3. I added a `_version.py` for versioning and we dynamically read the version in `setup.py`.

-----

This PR cleans up the `MegaBlocks` installation process. Specifically, it
1. Adds a `pyproject.toml` file
4. Updates the `README.md` installation instructions
5. Ignores more files in `.gitignore`
6. Adds missing packages to `setup.py`, i.e. `torch` and `numpy`
7. Adds missing packages for testing to `megablocks[dev]`, i.e. `coverage`, `pytest`, `pytest_codeblocks`, `pytest-cov`, `pre-commit`. 
8. Gracefully handles missing `torch` while running `setup.py`
9. Fix style in `setup.py`: replade `'` with `"` and add license header
10. Improve `long_description` by parsing `README.md` better (taken from `composer`)
11. `MegaBlocks` depends on `stk`. `stk`'s `setup.py` unnecessarily imports torch which causes issues with installing `stk` before we have `torch` installed. We now install my [fork](https://github.com/eitanturok/stk) of `stk` that fixes this issue while waiting for @tgale96 to merge my [PR](https://github.com/stanford-futuredata/stk/pull/13) fixing this in `stk`.

### CODE REVIEW

File: setup.py
Comment (b-chu) [inline setup.py:95]: Remove extra README

File: setup.py
Comment (b-chu) [inline setup.py:53]: Follow Composer style

File: setup.py
Comment (b-chu) [inline setup.py:29]: Why are we versioning on your fork?

File: pyproject.toml
Comment (b-chu) [inline pyproject.toml:1]: Base off composer and add C++ changes as necesary

File: setup.py
Comment (b-chu) [inline setup.py:11]: Unused

File: .gitignore
Comment (b-chu) [inline .gitignore:1]: Where is this taken from?

**Review by b-chu — COMMENTED** [review]
Thanks for the PR!

File: .gitignore
Comment (eitanturok) [reply .gitignore:1]: Took Composer's .gitignore

File: setup.py
Comment (eitanturok) [reply setup.py]: Removed.

File: setup.py
Comment (eitanturok) [reply setup.py]: Removed these comments.

File: setup.py
Comment (eitanturok) [reply setup.py]: Removed.

File: pyproject.toml
Comment (eitanturok) [reply pyproject.toml:1]: Addressed.

File: pyproject.toml
Comment (eitanturok) [reply pyproject.toml]: Removed irrelevant code snippets.

**Comment by eitanturok** [comment]
I addressed all of your comments @b-chu except for the one about using my fork of `stk`.

My fork enables us to set up github actions for MegaBlocks, something we do not need to worry about right now. So I'm going to just use the regular `stk` version so we can merge this.

Thoughts?

File: megablocks/__init__.py
Comment (eitanturok) [reply megablocks/__init__.py:7]: This docstring is taken from composer. I think we should keep things consistent with the composer style.

File: setup.py
Comment (eitanturok) [reply setup.py:139]: I'm waiting for Kobe to get back me about updating this.

**Review by b-chu — APPROVED** [review]

**Comment by eitanturok** [comment]
Are we ready to merge? Any other reviewers we should notify?

File: README.md
Comment (dakinggg) [inline README.md:25]:
```suggestion
To contribute to MegaBlocks, install `megablocks[dev]`. Run `pre-commit install` to configure the [pre-commit](https://pre-commit.com/) hook and then run `pre-commit run` before each commit to automatically format the code.
```

**Comment by karan6181** [comment]
Tagging @mvpatel2000 for the review before we merge this PR.

File: setup.py
Comment (dakinggg) [inline setup.py:20]:
```suggestion
        "No module named 'torch'. `torch` is required to install megablocks."
```

**Review by dakinggg — APPROVED** [review]
LGTM, but wanted to note that since this PR includes a `pyproject.toml`, I think it enables build isolation which may make the package more difficult to install. Since the torch dep is specified in pyproject.toml it should be ok, but upgrading to latest torch versions will need to be done with care.

File: setup.py
Comment (eitanturok) [reply setup.py:20]: Resolved.

File: README.md
Comment (eitanturok) [reply README.md]: Addressed.

File: megablocks/__init__.py
Comment (karan6181) [inline megablocks/__init__.py:1]: Do you have a plan to include the [header file](https://github.com/mosaicml/llm-foundry/blob/main/.pre-commit/FILE_HEADER) either in this PR or next? This will automatically add the copyright header when running the pre-commit.

Comment (eitanturok) [reply megablocks/__init__.py:1]: I'm doing this in a follow up PR.

**Comment by eitanturok** [comment]
Currently using custom forks of packages to resolve installation issues:
* https://git@github.com/eitanturok/stk.git for stanford-stk
* git+https://git@github.com/eitanturok/grouped_gemm.git for grouped_gemm

Blocked until these are merged. @dakinggg

**Comment by eitanturok** [comment]
Closing this PR since #127 changed a lot of things.

**Pull request closed without merging**
